### ISSUE TITLE
openapi-typescript 7.0.0 generates an additional `| undefined` to object types

### ISSUE THREAD

**Issue by nofalx** [openapi-ts/openapi-typescript#1727]
**Description**

In the new version the typescript generated for `object` types has an addition  `| undefined` that was not there before

```
Paste full error message
```

| Name                 | Version                      |
| :------------------- | :--------------------------- |
| `openapi-typescript` | `7.0.0`                      |
| Node.js              | `20.14.0`                      |
| OS + version         | `Arch Linux` |

**Reproduction & Expected result**

for `"openapi-typescript": "6.7.6"` I get the below output

```js
        /** CalculationRecordOutSchema */
        CalculationRecordOutSchema: {
            /** Channel Based Weights */
            channel_based_weights: {
                [key: string]: components['schemas']['ChannelWeightInfo'];
            };
        };
```

but after upgrading to `7.0.0` I get the below

```js
        /** CalculationRecordOutSchema */
        CalculationRecordOutSchema: {
            channel_based_weights: {
                [key: string]: components["schemas"]["ChannelWeightInfo"] | undefined;
            };
        };
```

My Schema is like this

```js
...
 "CalculationRecordOutSchema": {
        "properties": {
          "channel_based_weights": {
            "additionalProperties": {
              "$ref": "#/components/schemas/ChannelWeightInfo"
            },
            "title": "Channel Based Weights",
            "type": "object"
          },
        },
        "required": [
          "channel_based_weights",
        ],
        "title": "CalculationRecordOutSchema",
        "type": "object"
      },
... // #/components/schemas/ChannelWeightInfo
      "ChannelWeightInfo": {
        "properties": {
          "id": {
            "title": "Id",
            "type": "integer"
          },
          "name": {
            "title": "Name",
            "type": "string"
          },
          "channel_weight": {
            "title": "Channel Weight",
            "type": "number"
          },
          "assessment_weight": {
            "title": "Assessment Weight",
            "type": "number"
          },
          "external_sources": {
            "additionalProperties": {
              "$ref": "#/components/schemas/ExternalSourcesDetail"
            },
            "title": "External Sources",
            "type": "object"
          }
        },
        "required": [
          "id",
          "name",
          "channel_weight",
          "assessment_weight",
          "external_sources"
        ],
        "title": "ChannelWeightInfo",
        "type": "object"
      },
```

**Checklist**

- [ ] I’m willing to open a PR (see [CONTRIBUTING.md](https://github.com/openapi-ts/openapi-typescript/blob/main/packages/openapi-typescript/CONTRIBUTING.md))

**Comment by nofalx** [comment]
The issues seems to only happen when using `[key: string]`

**Comment by nofalx** [comment]
After looking at the schema more it seems to happen when `"type": "object"` and having `"additionalProperties": {...}` at the same time

**Comment by nofalx** [comment]
Makes sense, for example

```
interface User {
  name: string;
  [key: string]: ContactInfo | undefined;
}
```

Even if a unknown key value seems to fit the `ContactInfo` structure, it doesn't mean that every key will always be there in every `User` object. 

We need to make sure these keys are actually present before using them.

**Comment by Zenor27** [comment]
I strongly disagree, Array or Object access return type should be `T | undefined` not because the type structure ensure that, but because you used the correct Typescript compiler option: `noUncheckedIndexedAccess: true`.

**Comment by shortercode** [comment]
I have experimented with this in the `Dictionary` type I use in most of my code.

```ts
type Dictionary<T = unknown> = Record<string, T | undefined>
```

But I came to the conclusion that including the `undefined` was a mistake, because it changed all values ( even existing ones ) to a union with `undefined` and required lots of frustrating runtime checks that didn't do anything. For example

```ts
const dict: Dictionary<number> = { a: 1, b: 2 }

let counter = 0
for (const value of Object.values(dict)) {
  // can't increment counter because value is incorrectly `number | undefined`
  counter += value 
}
```

Moving from 6.x to 7.x has actually added a number of frustrating non-errors to my codebase for any dictionary like values which require either dangerous casts or pointless filtering to circumvent. ( yes I checked and already have checks in place for each error ).

I know that property access gives false positives on existence in indexed types, but as @Zenor27 says the correct course is to enable `noUncheckedIndexedAccess: true` in the tsconfig. Yes it's frustrating to use, but it enforces checks in places where a value can be undefined, without polluting the type with Nullish values. The fact that this is disabled by default is a TypeScript DX problem, not a bug in your library.

**Comment by shkar-sulysoft** [comment]
same issue here generate undefined for every open API specification

**Comment by drwpow** [comment]
Given the feedback on this, I think it’s probably better to reverse-course and match v6 implementation. I think the reasoning I gave may have been from even before v6, and the rewrite may have accidentally regressed this behavior.

I’ll open a PR to reverse this behavior in a patch, and let TypeScript do the handling.

This gets really close to a “breaking change,” but given the specific nature of how this works, removing `| undefined` shouldn’t actually break anything as far as I’m aware of; it should just fix the opened issues. However, if we did the opposite—add that `| undefined`, that would be a breaking change.
